### PR TITLE
mastodon: generate and read yarn hash from dependencies dir

### PR DESCRIPTION
**Pull request by meatballhat**
## Description of changes

:warning: This is my attempt at picking up the work from #174202 but without the "extra secrets" bits since those are now handled via `extraEnvFiles`. I attempted to incorporate feedback from #174202 comments 🤞🏼

~I'm not convinced that this is actually doing what I want it to do since running the `mastodon.updateScript` when pointing my flake at this branch doesn't appear to be using the altered `update.sh` (???) (see [flake with updates in branch](https://gitlab.com/tootcat/nixos-config/-/merge_requests/38)).~ Feedback/help very much wanted 🙇🏼 

cc @ashkitten 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/mastodon/default.nix [hunk 1]
```diff
@@ -39,11 +40,11 @@ stdenv.mkDerivation rec {
 
   mastodonModules = stdenv.mkDerivation {
     pname = "${pname}-modules";
-    inherit src version;
+    inherit src version yarnSha256;
 
     yarnOfflineCache = fetchYarnDeps {
       yarnLock = "${src}/yarn.lock";
-      sha256 = "sha256-e3rl/WuKXaUdeDEYvo1sSubuIwtBjkbguCYdAijwXOA=";
+      sha256 = yarnSha256;
```
Comment (dotlambda) [inline pkgs/servers/mastodon/default.nix:47]:
```suggestion
      hash = yarnHash;
```

File: pkgs/servers/mastodon/default.nix [hunk 2]
```diff
@@ -7,7 +7,8 @@
 , pname ? "mastodon"
 , version ? import ./version.nix
 , srcOverride ? null
-, dependenciesDir ? ./.  # Should contain gemset.nix, yarn.nix and package.json.
+, dependenciesDir ? ./.  # Expected to contain gemset.nix
+, yarnSha256 ? import ./yarn-sha256.nix
```
Comment (dotlambda) [inline pkgs/servers/mastodon/default.nix:11]:
```suggestion
, yarnHash ? import ./yarn-hash.nix
```

**Comment by meatballhat** [comment]
following up with more "yarn sha256" ➡️ "yarn hash" shortly

File: pkgs/build-support/node/fetch-yarn-deps/index.js [hunk 3]
```diff
@@ -37,7 +37,9 @@ const downloadFileHttps = (fileName, url, expectedHash, hashType = 'sha1') => {
 			res.on('end', () => {
 				file.close()
 				const h = hash.read()
-				if (h != expectedHash) return reject(new Error(`hash mismatch, expected ${expectedHash}, got ${h}`))
+				if (expectedHash === undefined){
```
Comment (erictapen) [inline pkgs/build-support/node/fetch-yarn-deps/index.js:40]: At first glance this looks fine to include in this PR, but could you put it in a separate commit and explain the motivation in the message? This way the change could be easily reverted if it would end up causing trouble.

File: pkgs/servers/mastodon/yarn-hash.nix [hunk 4]
```diff
@@ -0,0 +1 @@
+"sha256-e3rl/WuKXaUdeDEYvo1sSubuIwtBjkbguCYdAijwXOA="
```
Comment (erictapen) [inline pkgs/servers/mastodon/yarn-hash.nix:1]: When I rerun the update script, it actually writes here :+1:
It comes up with `1q2wy0l04796p3h4d3j11ciyxrjadj6vw61ig0fsapcadgyyaykv` though, without the `sha256` prefix.

File: pkgs/build-support/node/fetch-yarn-deps/index.js (hunk 3 above)
Comment (meatballhat) [reply pkgs/build-support/node/fetch-yarn-deps/index.js:40]: @erictapen thank you for calling this out. I'll get the commit separated 👍🏼

File: pkgs/servers/mastodon/yarn-hash.nix (hunk 4 above)
Comment (meatballhat) [reply pkgs/servers/mastodon/yarn-hash.nix:1]: When I run the update script _within_ nixpkgs via `nix-shell ./maintainers/scripts/update.nix --argstr package mastodon` I also see this behavior 👍🏼 I'll update the `yarn-hash.nix` file accordingly.

What I can't figure out, and I'd been chalking up to my being very new to nix, is why [this usage](https://gitlab.com/tootcat/nixos-config/-/blob/79890506fb038a7727d6b9e4e2538314ab096411/update.sh#L29) seems to be running the update script from a different rev of nixpkgs (without the `yarn-hash.nix` generation) even though the [flake specifies this branch](https://gitlab.com/tootcat/nixos-config/-/blob/79890506fb038a7727d6b9e4e2538314ab096411/flake.nix#L5). What do you think?

File: pkgs/build-support/node/fetch-yarn-deps/index.js (hunk 3 above)
Comment (erictapen) [reply pkgs/build-support/node/fetch-yarn-deps/index.js:40]: Are you sure you also use the latest revision? Your [flake.lock](https://gitlab.com/tootcat/nixos-config/-/blob/79890506fb038a7727d6b9e4e2538314ab096411/flake.lock) doesn't point to the current commit. You might need to run `nix flake update --commit-lock-file` first.

File: pkgs/servers/mastodon/yarn-hash.nix (hunk 4 above)
Comment (erictapen) [reply pkgs/servers/mastodon/yarn-hash.nix:1]: Are you sure you also use the latest revision? Your [flake.lock](https://gitlab.com/tootcat/nixos-config/-/blob/79890506fb038a7727d6b9e4e2538314ab096411/flake.lock) doesn't point to the current commit. You might need to run nix flake update --commit-lock-file first.

Comment (meatballhat) [reply pkgs/servers/mastodon/yarn-hash.nix:1]: OK, so I ran `nix flake update --commit-lock-file` which *did* update the [flake.lock](https://gitlab.com/tootcat/nixos-config/-/blob/3ee2d4510db0bc1d9a866e506584493d8914c454/flake.lock) but when running the update script I'm still seeing behavior *without* the generation of `yarn-hash.nix`, like the `mastodon.updateScript` in use is from a different rev 🤔

Comment (erictapen) [reply pkgs/servers/mastodon/yarn-hash.nix]: Ah, now I see it. Your `update.sh` uses `nix-shell` in the shebang, which uses the channel mechanism to aquire Nixpkgs. That is the "old" way of doing things and runs completely in parallel to flakes.

Comment (meatballhat) [reply pkgs/servers/mastodon/yarn-hash.nix]: > When I run the update script _within_ nixpkgs via `nix-shell ./maintainers/scripts/update.nix --argstr package mastodon` I also see this behavior 👍🏼 I'll update the `yarn-hash.nix` file accordingly.

I take this back. Running `nix-shell ./maintainers/scripts/update.nix --argstr package mastodon` does not work for me as-is.

<details>
  <summary>script output</summary>

```
Going to be running update for following packages:
 - mastodon-4.1.7

Press Enter key to continue...

Running update for:
 - mastodon-4.1.7: UPDATING ...
 - mastodon-4.1.7: ERROR

--- SHOWING ERROR LOG FOR mastodon-4.1.7 ----------------------

error: hash 'null' does not include a type, nor is the type otherwise known from context
(use '--show-trace' to show detailed location information)


--- SHOWING ERROR LOG FOR mastodon-4.1.7 ----------------------
The update script for mastodon-4.1.7 failed with exit code 1
```

</details>

<details>
  <summary>mastodon.log</summary>

```
error: hash 'null' does not include a type, nor is the type otherwise known from context
(use '--show-trace' to show detailed location information)
```

</details>

<details>
  <summary>git diff after script fail</summary>

```diff
diff --git a/pkgs/servers/mastodon/source.nix b/pkgs/servers/mastodon/source.nix
index e0920eadf5f2..ab279088ab4d 100644
--- a/pkgs/servers/mastodon/source.nix
+++ b/pkgs/servers/mastodon/source.nix
@@ -4,7 +4,7 @@
     owner = "mastodon";
     repo = "mastodon";
     rev = "v4.1.7";
-    hash = "sha256-afGv7/Cl7bxbHvLXrtpZPlG4oJeL7LgQQKXBBt3LRSo=";
+    hash = "null";
   };
 in applyPatches {
   inherit src;
```

</details>

If I modify `pkgs/servers/mastodon/update.sh` like this, I get a successful update:

<details>
  <summary><code>pkgs/servers/mastodon/update.sh</code> diff</summary>

```diff
diff --git a/pkgs/servers/mastodon/update.sh b/pkgs/servers/mastodon/update.sh
index 1266770a4364..36669c2885cc 100755
--- a/pkgs/servers/mastodon/update.sh
+++ b/pkgs/servers/mastodon/update.sh
@@ -77,7 +77,7 @@ trap cleanup EXIT

 echo "Fetching source code $REVISION"
 JSON=$(nix-prefetch-github "$OWNER" "$REPO" --rev "$REVISION" 2> $WORK_DIR/nix-prefetch-git.out)
-HASH=$(echo "$JSON" | jq -r .hash)
+HASH=$(echo "$JSON" | jq -r .sha256)

 echo "Creating version.nix"
 echo "\"$VERSION\"" | sed 's/^"v/"/' > version.nix
@@ -89,7 +89,7 @@ cat > source.nix << EOF
     owner = "$OWNER";
     repo = "$REPO";
     rev = "$REVISION";
-    hash = "$HASH";
+    hash = "sha256-$HASH";
   };
 in applyPatches {
   inherit src;
```

</details>

Comment (meatballhat) [reply pkgs/servers/mastodon/yarn-hash.nix]: > Ah, now I see it. Your `update.sh` uses `nix-shell` in the shebang, which uses the channel mechanism to aquire Nixpkgs. That is the "old" way of doing things and runs completely in parallel to flakes.

Oh! TIL ❤️ Do you know offhand where I should be looking to learn how to change the `update.sh` to use flakes correctly? (fwiw I inherited this code 🤷🏼)

**Comment by meatballhat** [comment]
I'll be reworking the yarn hash to work more like [this](https://github.com/NixOS/nixpkgs/blob/407c4acda970eacd51f25fc073e1b8585aa4d902/pkgs/tools/admin/pgadmin/update.sh#L97-L100)

File: pkgs/servers/mastodon/yarn-hash.nix (hunk 4 above)
Comment (erictapen) [reply pkgs/servers/mastodon/yarn-hash.nix:1]: Either put the dependencies into a flake `devShells` attribute and run the script within that shell, or you build it with Nix (e.g. with `pkgs.writeShellScriptBin "update" "your script content"`) as a flake output with all packages called explicitly (e.g. `${pkgs.yarn2nix}/bin/yarn2nix --bla`) and then run that via `nix run .#update`.

File: pkgs/servers/mastodon/default.nix [hunk 5]
```diff
@@ -39,11 +40,11 @@ stdenv.mkDerivation rec {
 
   mastodonModules = stdenv.mkDerivation {
     pname = "${pname}-modules";
-    inherit src version;
+    inherit src version yarnHash;
 
     yarnOfflineCache = fetchYarnDeps {
       yarnLock = "${src}/yarn.lock";
-      sha256 = "sha256-e3rl/WuKXaUdeDEYvo1sSubuIwtBjkbguCYdAijwXOA=";
+      hash = yarnHash;
```
[the remark this answers is not in the document]
Comment (erictapen) [reply pkgs/servers/mastodon/default.nix:47]: I'd rather not add more sed logic to Nixpkgs, I think it's a good solution to just have it in a separate file, like the version and source.

File: pkgs/servers/mastodon/default.nix [hunk 6]
```diff
@@ -39,11 +40,11 @@ stdenv.mkDerivation rec {
 
   mastodonModules = stdenv.mkDerivation {
     pname = "${pname}-modules";
-    inherit src version;
+    inherit src version yarnHash;
```
[the remark this answers is not in the document]
Comment (meatballhat) [reply pkgs/servers/mastodon/default.nix]: We don't! Thank you 🙇🏼

**Review by erictapen — APPROVED** [review]
Thanks!

File: pkgs/servers/mastodon/default.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/servers/mastodon/default.nix:47]: Wouldn't it then be a better idea to move it everything into one file?

[the remark this answers is not in the document]
Comment (meatballhat) [reply pkgs/servers/mastodon/default.nix:47]: @SuperSandro2000 imho no, because different tools are generating these strings and it's much simpler to assume that the file can be completely overwritten

File: pkgs/servers/mastodon/update.sh [hunk 7]
```diff
@@ -77,7 +77,8 @@ trap cleanup EXIT
 
 echo "Fetching source code $REVISION"
 JSON=$(nix-prefetch-github "$OWNER" "$REPO" --rev "$REVISION" 2> $WORK_DIR/nix-prefetch-git.out)
-HASH=$(echo "$JSON" | jq -r .hash)
+HASH="$(echo "$JSON" | jq -r .sha256)"
+HASH="$(nix hash to-sri --type sha256 "$HASH")"
```
Comment (SuperSandro2000) [inline pkgs/servers/mastodon/update.sh:81]: This breaks the update script if you are on nixos-unstable. Going to revert it.

Comment (erictapen) [reply pkgs/servers/mastodon/update.sh:81]: Weird, I thought I tested this. Does this help? https://github.com/NixOS/nixpkgs/pull/257810
